includes/class-wcsatt-product.php

```diff
@@ -0,0 +1,856 @@
+<?php
+/**
+ * WCS_ATT_Product API
+ *
+ * @author   SomewhereWarm <info@somewherewarm.gr>
+ * @package  WooCommerce Subscribe All the Things
+ * @since    2.0.0
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * API for working with subscription-enabled product objects.
+ *
+ * @class    WCS_ATT_Product
+ * @version  2.0.0
+ */
+class WCS_ATT_Product {
+
+	public static function init() {
+
+		// Allow WCS to recognize any product as a subscription.
+		add_filter( 'woocommerce_is_subscription', array( __CLASS__, 'filter_is_subscription' ), 10, 3 );
+
+		// Delete object meta in use by the application layer.
+		add_action( 'woocommerce_before_product_object_save', array( __CLASS__, 'delete_reserved_meta' ) );
+
+		// Make product prices scheme-dependent.
+		WCS_ATT_Scheme_Prices::add_price_filters();
+		WCS_ATT_Scheme_Prices::add_price_html_filters();
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Conditionals
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Determines if a subscription scheme is set on the product object.
+	 *
+	 * @param  WC_Product  $product  Product object to check.
+	 * @return boolean               Result of check.
+	 */
+	public static function is_subscription( $product ) {
+		$active_scheme_key = self::get_subscription_scheme( $product );
+		return ! empty( $active_scheme_key );
+	}
+
+	/**
+	 * Determines if the product can be purchased on a recurring basis.
+	 *
+	 * @param  WC_Product  $product  Product object to check.
+	 * @return boolean               Result of check.
+	 */
+	public static function has_subscriptions( $product ) {
+		return sizeof( self::get_subscription_schemes( $product ) ) > 0;
+	}
+
+	/**
+	 * Determines if the product is purchasable on a recurring basis only.
+	 *
+	 * @param  WC_Product  $product  Product object to check.
+	 * @return boolean               Result of check.
+	 */
+	public static function has_forced_subscription( $product ) {
+
+		if ( '' === ( $forced = self::get_product_property( $product, 'has_forced_subscription' ) ) && self::has_subscriptions( $product ) ) {
+
+			$forced = WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ? $product->get_meta( '_wcsatt_force_subscription', true ) : get_post_meta( WCS_ATT_Core_Compatibility::get_id( $product ), '_wcsatt_force_subscription', true );
```

Comment:
In most cases, you will notice that the raw data obtained via `get_meta` or `get_post_meta` requires some kind of transformation -- then the approach followed here is to store/cache it on the object using `set_product_property`.

Two examples:

1. "Raw" subscription scheme data is used to create scheme objects once. The created objects are stored/cached on the product as "runtime meta".

2. "Raw" subscription scheme data only includes the schemes defined at product level, but cart-level schemes can be appended to product objects in the cart after loading cart session data. 

So `get/set_product_property` is basically used as a "runtime meta" getter/setter, which does some back-compat handling.

Also see https://github.com/Prospress/woocommerce-subscribe-all-the-things/pull/136#discussion_r123648135